### PR TITLE
[parallel] Add projection support to hpx::is_sorted, hpx::is_sorted_until, and hpx::is_partitioned CPOs

### PR DESCRIPTION
**Pull request by aneek22112007-tech**
## Summary

This PR adds projection support to the `hpx::is_sorted`, `hpx::is_sorted_until`, and `hpx::is_partitioned` CPOs. The underlying parallel implementations already accepted projections via the `Proj` template parameter, but the user-facing `hpx::` namespace CPOs lacked the necessary `tag_fallback_invoke` overloads to forward them.

### Algorithm Headers

- **`is_sorted.hpp`** — Added `tag_fallback_invoke` overloads for `hpx::is_sorted_t` and `hpx::is_sorted_until_t` that accept an optional `Proj` parameter with sensible defaults.
- **`is_partitioned.hpp`**:
  - Added `tag_fallback_invoke` overloads for `hpx::is_partitioned_t` that forward projections to the internal implementation.
  - Renamed `partition_status` enum values from `true_`/`false_` to `is_partitioned`/`is_not_partitioned` for clarity. Added explicit underlying type `std::uint8_t`.
  - Removed the unnecessary `local_policy` type alias; now uses `util::const_loop_n<std::decay_t<ExPolicy>>` directly, matching the pattern used in `is_sorted`.

### Test Suite

- **`is_sorted_projection.cpp`** *(new)* — Comprehensive projection tests using `std::pair` and custom lambda projections, covering sequential, `par`, and `par(task)` execution policies, plus edge cases for empty/single-element ranges.
- **`CMakeLists.txt`** — Registered the new test target.

## Checklist

- [x] I have fixed a bug and added a regression test.
- [x] Addressed all maintainer review comments (enum naming, local_policy removal).
- [x] Resolved merge conflicts with upstream `master` (adopted upstream `const_loop_n` and sentinel-based boundary check for `is_sorted`).

### CODE REVIEW

**Review by hkaiser — COMMENTED** [review]
Thanks for your work on this.

We generally try to stay as close to the C++ Standard as possible. For this reason I don't think that we should go ahead with adding projections to the algorithms in `namespace hpx`.

Does this PR add anything in addition to the API extensions that would be worth keeping?

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp [hunk 1]
```diff
@@ -215,29 +195,49 @@ namespace hpx::parallel {
                     -> intermediate_result_t {
                     bool fst_bool = HPX_INVOKE(pred_projected, *part_begin);
                     if (part_count == 1)
-                        return fst_bool;
+                        return fst_bool ? 1 : 0;
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp]: Please use an `enum class` for this.

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp [hunk 2]
```diff
@@ -283,6 +289,48 @@ namespace hpx {
...
+            )
+        // clang-format on
+        friend decltype(auto) tag_fallback_invoke(hpx::is_partitioned_t,
+            ExPolicy&& policy, FwdIter first, FwdIter last, Pred pred,
+            Proj proj)
+        {
+            return hpx::parallel::detail::is_partitioned<FwdIter, FwdIter>()
+                .call(HPX_FORWARD(ExPolicy, policy), first, last,
+                    HPX_MOVE(pred), HPX_MOVE(proj));
+        }
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp:313]: Please revert the API changes.

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp [hunk 3]
```diff
@@ -11,233 +11,20 @@
...
-    ///           element. If the sequence has less than two elements or the
-    ///           sequence is sorted, last is returned.
-    ///
-    template <typename ExPolicy, typename FwdIter,
-        typename Pred = hpx::parallel::detail::less>
-    typename hpx::parallel::util::detail::algorithm_result<ExPolicy,
-        FwdIter>::type
-    is_sorted_until(
-        ExPolicy&& policy, FwdIter first, FwdIter last, Pred&& pred = Pred());
-}    // namespace hpx
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp:224]: WHy have you removed the documentation?

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp [hunk 4]
```diff
@@ -11,233 +11,20 @@
...
-#else
-
 #include <hpx/config.hpp>
-#include <hpx/modules/coroutines.hpp>
-#include <hpx/modules/executors.hpp>
-#include <hpx/modules/functional.hpp>
-#include <hpx/modules/iterator_support.hpp>
-#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+
+#include <hpx/algorithms/traits/projected.hpp>
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp]: Please always use the generated module headers to avoid problems with C++ modules.

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp [hunk 5]
```diff
@@ -304,10 +91,10 @@ namespace hpx::parallel {
                         FwdIter_ part_begin, std::size_t part_size) mutable
                     -> intermediate_result_t {
                     FwdIter_ trail = part_begin++;
-                    util::loop_n<std::decay_t<ExPolicy>>(part_begin,
+                    util::loop_n<hpx::execution::sequenced_policy>(part_begin,
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp]: This may lose context (such like vectorization, etc.)

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp [hunk 6]
```diff
@@ -329,7 +116,7 @@ namespace hpx::parallel {
 
                 auto f2 = [](auto&& results) {
                     return std::all_of(hpx::util::begin(results),
-                        hpx::util::end(results), hpx::functional::unwrap{});
+                        hpx::util::end(results), [](auto x) { return x; });
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp]: What is the rationale of this change, please explain. Did some tests fail?

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp [hunk 7]
```diff
@@ -433,18 +219,15 @@ namespace hpx::parallel {
...
+                    {
+                        return detail::advance(first, cancelled_at);
+                    }
+                    return detail::advance(first, count);
                 };
 
-                using partitioner_type =
-                    util::partitioner<policy_type, FwdIter_, void>;
-                return partitioner_type::call_with_index(
+                return util::partitioner<ExPolicy, FwdIter_>::call_with_index(
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp]: This will again lose context.

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp [hunk 8]
```diff
@@ -496,6 +279,51 @@ namespace hpx {
...
+            )
+        // clang-format on
+        friend decltype(auto) tag_fallback_invoke(hpx::is_sorted_t,
+            ExPolicy&& policy, FwdIter first, FwdIter last, Pred pred,
+            Proj proj)
+        {
+            return hpx::parallel::detail::is_sorted<FwdIter, FwdIter>().call(
+                HPX_FORWARD(ExPolicy, policy), first, last, HPX_MOVE(pred),
+                HPX_MOVE(proj));
+        }
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp:498]: Please roll back all API changes.

File: libs/core/algorithms/include/hpx/parallel/container_algorithms/is_partitioned.hpp [hunk 9]
```diff
@@ -325,14 +318,14 @@ namespace hpx::ranges {
 
             return hpx::parallel::detail::is_partitioned<iterator_type,
                 iterator_type>()
-                .call(hpx::execution::seq, std::begin(rng), std::end(rng),
-                    HPX_MOVE(pred), HPX_MOVE(proj));
+                .call(hpx::execution::seq, hpx::util::begin(rng),
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/container_algorithms/is_partitioned.hpp:328]: What's the rationale of this change? I'd have expected to see changes the other way around (i.e., moving towards the standard, not away from it).

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp (hunk 1 above)
Comment (aneek22112007-tech) [reply libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp]: Fixed. I've introduced enum class partition_status to handle the different partitioning states clearly.

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp [hunk 10]
```diff
@@ -136,36 +139,22 @@ namespace hpx::parallel {
     ////////////////////////////////////////////////////////////////////////////
     // is_partitioned
     namespace detail {
+        enum class partition_status : std::uint8_t
+        {
+            false_ = 0,
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp]: Could you come up with slightly more descriptive names, please? What is `false_`, what is `true_`? Does it refer to whether a given chunk is partitioned? Why not call it `is_partitioned` and `is_not_partitioned` in this case?

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp [hunk 11]
```diff
@@ -208,36 +196,68 @@ namespace hpx::parallel {
...
+                        return fst_bool ? partition_status::true_ :
+                                          partition_status::false_;
+
+                    bool is_mixed = false;
+
+                    using local_policy = std::conditional_t<
+                        hpx::is_unsequenced_execution_policy_v<
+                            std::decay_t<ExPolicy>>,
+                        hpx::execution::unsequenced_policy,
+                        hpx::execution::sequenced_policy>;
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp]: Why is this needed?

**Comment by aneek22112007-tech** [comment]
Hi @hkaiser, I've addressed both review comments:

1. **Renamed partition_status enum values**: Changed `false_` → `is_not_partitioned` and `true_` → `is_partitioned` for clarity, and added the underlying type `std::uint8_t`.

2. **Removed unnecessary `local_policy`**: Replaced the `std::conditional_t` type alias and `util::loop_n<local_policy>` with `util::const_loop_n<std::decay_t<ExPolicy>>` directly, matching the pattern used in the upstream `is_sorted` implementation. Also typed the callback iterator as `Iter_` instead of `auto`.

**Review by hkaiser — DISMISSED** [review]
LGTM, thanks!

**Comment by hkaiser** [comment]
@aneek22112007-tech could you please fix the compilation issues reported (https://github.com/TheHPXProject/hpx/actions/runs/25022403344/job/73289688243?pr=7220, https://github.com/TheHPXProject/hpx/actions/runs/25022403317/job/73285788254?pr=7220)?

**Comment by hkaiser** [comment]
@aneek22112007-tech ping? the PR still exposes compilation errors. Could you please make sure to fix those before commiting more things?

**Comment by hkaiser** [comment]
Inspect reports:
```
/libs/core/algorithms/tests/unit/algorithms/is_sorted_projection.cpp: *I* missing #include (string) for symbol std::string on line 144
```

File: libs/core/algorithms/include/hpx/algorithms/traits/projected.hpp [hunk 12]
```diff
@@ -20,15 +20,15 @@
 namespace hpx::traits {
 
     ///////////////////////////////////////////////////////////////////////////
-    HPX_CXX_CORE_EXPORT template <typename T, typename Enable = void>
+    template <typename T, typename Enable = void>
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/algorithms/traits/projected.hpp]: I don't think it is a good idea to remove the C++ module annotations (and the compilation errors we see confirm that).

Comment (aneek22112007-tech) [reply libs/core/algorithms/include/hpx/algorithms/traits/projected.hpp]: Thanks for the feedback! I have restored the HPX_CXX_CORE_EXPORT annotations across projected.hpp, is_sorted.hpp, and is_partitioned.hpp to ensure proper C++ module visibility. I also added the missing <string> include to the is_sorted_projection test to resolve the hpxinspect report. All files have been re-formatted using clang-format-20 to match the CI environment.

File: libs/core/algorithms/include/hpx/algorithms/traits/projected_range.hpp [hunk 13]
```diff
@@ -15,13 +15,12 @@
 namespace hpx::parallel::traits {
 
     ///////////////////////////////////////////////////////////////////////////
-    HPX_CXX_CORE_EXPORT template <typename F, typename Rng,
-        typename Enable = void>
+    template <typename F, typename Rng, typename Enable = void>
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/algorithms/traits/projected_range.hpp]: This probably needs the `HPX_CXX_CORE_EXPORT` annotation as well (same is true for the other symbols in this file).

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp (hunk 2 above)
Comment (hkaiser) [reply libs/core/algorithms/include/hpx/parallel/algorithms/is_partitioned.hpp:313]: Will you revert these changes to the API, please?

File: libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp [hunk 14]
```diff
@@ -459,85 +459,99 @@ namespace hpx {
...
         // clang-format on
-        friend bool tag_fallback_invoke(
-            hpx::is_sorted_t, FwdIter first, FwdIter last, Pred pred = Pred())
+        friend bool tag_fallback_invoke(hpx::is_sorted_t, FwdIter first,
+            FwdIter last, Pred pred = Pred(), Proj proj = Proj())
         {
             return hpx::parallel::detail::is_sorted<FwdIter, FwdIter>().call(
                 hpx::execution::seq, first, last, HPX_MOVE(pred),
-                hpx::identity_v);
+                HPX_MOVE(proj));
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp]: Please revert the API changes as discussed.

Comment (aneek22112007-tech) [reply libs/core/algorithms/include/hpx/parallel/algorithms/is_sorted.hpp]: I have reverted the API changes to the hpx::is_sorted, hpx::is_sorted_until, and hpx::is_partitioned CPOs as requested. They now match their original signatures while still utilizing the updated internal parallel implementations (which maintain projection support for the hpx::ranges versions). I also restored the HPX_CXX_CORE_EXPORT annotations in projected_range.hpp and other trait headers. All tests have been updated and verified.

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!
